### PR TITLE
Ellipse apertures

### PR DESCRIPTION
**Pull request by jorgemarpa**
This PR adds the functionality to create an elliptic aperture. It uses first- and second-order moments to compute the flux distribution of the object and estimate the ellipse parameters (semi-major, semi-minor axes, and orientation angle). It follows the recipe described [here](https://astromatic.github.io/sextractor/Position.html#ellipse-iso-def).
The aperture mask selects pixels inside the fitted ellipse scaled by the parameter R. 
the PR adds:
* `MovingTPF._ellipse_aperture()`: a hidden method that computes the ellipse parameters and returns the aperture mask. Includes a plotting option to show moment values.
*  `utils.py`: compilation of utility methods, like `compute_moments()` from a 2D image and `inside_ellipse()` that uses the quadratic parametrization of an ellipse to select pixels inside the ellipse.
* `test_ellipse_aperture()`: test the new implementation
* `MovingTPF.create_aperture()`: now accepts `method="ellipse"`

### CODE REVIEW

File: tests/test_movingtpf.py
Comment (altuson) [inline tests/test_movingtpf.py:212]:
```suggestion
    Test ellipse aperture. The ellipse is calculated by computing the first- and second-order
```

Comment (altuson) [reply tests/test_movingtpf.py:212]: Not sure it's right to say fitted when the ellipse parameters are calculated from moments.

File: tests/test_movingtpf.py
Comment (altuson) [inline tests/test_movingtpf.py:214]:
```suggestion
    ellipse, x/y centroid and the ellipse parameters (semi-major/minor axis and rotation angle).
```

File: tests/test_movingtpf.py
Comment (altuson) [inline tests/test_movingtpf.py:223]:
```suggestion
    ellip_mask, params = test._ellipse_aperture(return_params=True)
```

File: tests/test_movingtpf.py
Comment (altuson) [inline tests/test_movingtpf.py:232]:
```suggestion
    assert params.shape == (test.time.shape[0], 5)
    assert np.isfinite(params).all()
    assert np.mean(params[:, 0] - test.ephemeris[:, 1]) < 0.2
    assert np.mean(params[:, 1] - test.ephemeris[:, 0]) < 0.2
```

File: tests/test_movingtpf.py
Comment (altuson) [inline tests/test_movingtpf.py:233]:
```suggestion
    assert ((params[:, 4] >= 0) & (params[:, 4] <= 360)).all()
```

File: src/tess_asteroids/utils.py
Comment (altuson) [inline src/tess_asteroids/utils.py:87]:
```suggestion
        row, col = np.mgrid[0 : flux.shape[1], 0 : flux.shape[2]]
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:395]:
```suggestion
            self.aperture_mask = self._ellipse_aperture(return_params=False, **kwargs)
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:492]:
```suggestion
        return_params: bool = False,
```
I think we should rename this `return_params` because it includes both centroids and ellipse parameters.

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:499]:
```suggestion
        Uses second-order moments of 2d flux distribution to compute ellipse parameters (cxx, cyy and cxy) and get an aperture mask with pixels inside the ellipse. The function can also optionally return the x/y centroids and the ellipse parameters (semi-major axis, A, semi-minor axis, B, and position angle, theta).
        Ref: https://astromatic.github.io/sextractor/Position.html#ellipse-iso-def
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:511]:
```suggestion
        return_params: boolean
            Return a ndarray with x/y centroids and ellipse parameters computed from first and second
            moments [X_cent, Y_cent, A, B, theta_deg].
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:520]:
```suggestion
        parameters: ndarray
            If `return_params`, also returns x/y centroid and ellipse parameters [X_cent, Y_cent, A, B, theta_deg] with shape (5, n_times).
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:568]:
```suggestion
        if return_params:
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:606]:
```suggestion
                y0=self.ephemeris[nt, 0],
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:609]:
```suggestion
        if return_params:
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:395]: Could we rename function `_create_ellipse_aperture` to be consistent with threshold aperture?

**Review by altuson — CHANGES_REQUESTED** [review]
Looks great! I've left some comments/ideas and fixed a couple of bugs. Let me know if you have any questions.

File: src/tess_asteroids/movingtpf.py
Comment (jorgemarpa) [reply src/tess_asteroids/movingtpf.py:498]: yes!

File: src/tess_asteroids/movingtpf.py
Comment (jorgemarpa) [reply src/tess_asteroids/movingtpf.py:524]: it should be 3.

File: src/tess_asteroids/movingtpf.py
Comment (jorgemarpa) [reply src/tess_asteroids/movingtpf.py:529]: I think is important to have diagnostic plots that show relevant model parameters. I'll add matplotlib as a dependency

File: pyproject.toml
Comment (jorgemarpa) [reply pyproject.toml:1]: done

File: src/tess_asteroids/movingtpf.py
Comment (jorgemarpa) [reply src/tess_asteroids/movingtpf.py:605]: This actually causes problems on bad frames with badly centroid estimations. I would like to add some iterative proccess to constrain the centroids better on those cases, but I think that would be in a further PR.

**Comment by jorgemarpa** [comment]
@altuson, I'm done with my 2nd revision.

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [reply src/tess_asteroids/movingtpf.py:564]: I can see you've now masked outliers in XY, but you've not masked zeros. Was that intentional?

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [reply src/tess_asteroids/movingtpf.py:565]: Please could you comment on this.

File: src/tess_asteroids/utils.py
Comment (altuson) [inline src/tess_asteroids/utils.py:13]:
```suggestion
    Pixels with distance <= R^2 from the center (x0, y0) are considered inside the ellipse.
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:631]:
```suggestion
            # subtraction has remaining artifacts.
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:518]:
```suggestion
            Generate a diagnostic plot with first- and second-order moments.
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:561]:
```suggestion
        # due to asteroid orbit projections, some tracks can show change in directions,
```

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:516]:
```suggestion
            Return a ndarray with x/y centroids and ellipse parameters computed from first- and second-order
            moments [X_cent, Y_cent, A, B, theta_deg].
```

**Review by altuson — CHANGES_REQUESTED** [review]
Finished second review. A couple of very minor comments.

**Comment by altuson** [comment]
Hi @jorgemarpa, ruff is currently failing so I can't merge the PR. Could you please look into this? Thanks!

File: src/tess_asteroids/movingtpf.py
Comment (altuson) [inline src/tess_asteroids/movingtpf.py:565]:
```suggestion
            mask = ~np.logical_and(Y2 == 0, np.logical_and(X2 == 0, XY == 0))
```
np.logical_and only takes two conditions, so you have to nest them.

**Review by altuson — CHANGES_REQUESTED** [review]

**Pull request merged**
